app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,157 @@
+      dependent = ar_object.public_send(association).public_send(:background_deletion_scope).take
+      dependent ? [hierarchy_association_string, *hierarchy_entries_for(dependent)] : []
+    when :has_one
+      # maximum of one associated so we never keep the original hierarchy entry
+      hierarchy_entries_for ar_object.public_send(association)
+    else
+      raise ArgumentError, "Cannot handle association #{ar_object}:#{association} type #{reflection.macro}"
     end
+  end
 
-    def destroy_later
-      reflection.many? ? destroy_has_many_association : destroy_has_one_association
-    end
+  def compatibility(_object, caller_worker_hierarchy = [], _background_destroy_method = 'destroy')
+    # maybe requeue first object from hierarchy would be adequate and uniqueness should deduplicate jobs
+    hierarchy_root = Array(caller_worker_hierarchy).first
+    return unless hierarchy_root
 
-    attr_reader :main_object, :reflection, :caller_worker_hierarchy
+    object_class, id = hierarchy_root.match(/Hierarchy-([a-zA-Z0-9_]+)-([\d*]+)/).captures
 
-    private
+    raise DoNotRetryError, "background deletion cannot handle #{hierarchy_root}" unless object_class && id
 
-    def destroy_has_many_association
-      main_object.public_send("#{reflection.name.to_s.singularize}_ids").each do |associated_object_id|
-        associated_object = reflection.class_name.constantize.new
-        associated_object.id = associated_object_id
-        delete_associated_object_later(associated_object)
-      end
-    rescue ActiveRecord::UnknownPrimaryKey => exception
-      Rails.logger.info "DeleteObjectHierarchyWorker#perform raised #{exception.class} with message #{exception.message}"
-    end
+    self.class.delete_later object_class.constantize.find(id.to_i)
+  end
 
-    def destroy_has_one_association
-      associated_object = main_object.public_send(reflection.name)
-      delete_associated_object_later(associated_object)
+  def associations_strings(ar_object, *associations)
+    associations = ar_object.class.background_deletion if associations.blank?
+    associations.map do |association|
+      "Association-#{ar_object.class}-#{ar_object.id}:#{association}"
     end
+  end
 
-    def delete_associated_object_later(associated_object)
-      association_delete_worker.perform_later(associated_object, caller_worker_hierarchy, reflection.background_destroy_method) if associated_object.try(:id)
-    end
+  delegate :info, to: 'Rails.logger'
 
-    def association_delete_worker
-      case reflection.class_name
-      when Account.name
-        DeleteAccountHierarchyWorker
-      when PaymentGatewaySetting.name
-        DeletePaymentSettingHierarchyWorker
-      else
-        DeleteObjectHierarchyWorker
-      end
+  def now
+    Process.clock_gettime(Process::CLOCK_MONOTONIC)
+  end
+
+  class DummyDestroyedByAssociationReflection
+    def initialize(foreign_key)
+      @foreign_key = foreign_key
     end
+
+    attr_reader :foreign_key
```

Comment:
Don't know what is it for but it feels wrong. Or maybe the class name is bad, but it looks like temporary/test code.

Reply:
It is a copy from the original code. No idea why it is defined like that. I think there was some comment about logging purposes. Maybe that's it, besides that some callbacks check whether deletion is by association or not.

Reply:
Do you think it will cause problems if we delete it?

Reply:
See the code comment added:
> callbacks logic differs between object destroyed by association, it is standalone if first job arg is itself

So if we delete an object standalone or as part of association, there is some callback logic differences. Whether they are important or not, I cannot say and it should be out of scope for this PR.

We can change the implementation of `DummyDestroyedByAssociationReflection` perhaps but it is beyond my scope for the PR unless there is an important reason to change it now instead of later in case anybody cares enough.